File: Data Structures & Algorithms/binary-tree-right-side-view/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        q = deque([root]) 

        while q:
            level_size = len(q)
            rightmost_node = None

            for _ in range(level_size):
                node = q.popleft()
                if node:
                    rightmost_node = node
                    q.append(node.left)
                    q.append(node.right)
            
            if rightmost_node:
                res.append(rightmost_node.val)

        return res

File: Data Structures & Algorithms/valid-binary-search-tree/submission-2.swift
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     var val: Int
 *     var left: TreeNode?
 *     var right: TreeNode?
 *     init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */

class Solution {
    func isValidBST(_ root: TreeNode?) -> Bool {
        guard let root else { return true }
        func isValid(_ node: TreeNode?, _ minVal: Int, _ maxVal: Int) -> Bool  {
            guard let node else { return true } 

            if node.val <= minVal || node.val >= maxVal {
                return false
            }

            let leftIsValid = isValid(node.left, minVal, node.val)
            let rightIsValid = isValid(node.right, node.val, maxVal)

            return leftIsValid && rightIsValid
        }

        return isValid(root, Int.min, Int.max)
    }
}
